features/323-login-id/design.md

```diff
@@ -0,0 +1,253 @@
+```
+If authentication succeed, both request would result in same user, with
+different principals.
+
+This addresses use case 10.
+
+
+## Use Principals in Auth API
+In a previous design (#294):
+- A list of all login IDs associated with current user would be returned along with authn response.
+- Login ID key used to authenticate current session would be returned in `LoginID-Key` header.
+- For passcodes (#293), `Passcode-ID` would be used.
+- User can update existing login ID
+
+This design has some flaws:
+- List of all Login IDs should not need to be present in authn response.
+  If required, a dedicated API should be used instead.
+- Other identity providers (e.g. OAuth, custom token) is not considered.
+
+In current design, the authn process returns a tuple (user, principal).
+
+A principal is associated with identity provider.
+To public, a principal is logically consisted of two parts:
+- **Principal Identity**: Defining features of the principal
+    - Password authentication: (login ID key, login ID)
+    - OAuth: (OAuth provider ID, origin principal ID)
+    - Custom token: (origin principal ID)
+- **Principal Metadata**: Additional information describing the principal
+    - Password authentication: Current none
+    - OAuth: (OAuth token, origin profile)
+    - Custom token: Current none
+Principal identity should be immutable, and principal metadata should be updated
+only through identity-provider specific APIs.
+
+
+To better model this design, the authn response should contains the principal instead.
+i.e.
+```json
+{
+    "access_token": "...",
+    "user": {
+        "id": "..."
+    },
+    "principal": {
+        "id": "...",
+        "realm": "teacher",
+        "email": "a@example.com",
+        "metadata": {}
+    }
+}
+```
+
+All auth APIs should be revised so that principal ID is used when appropiate.
+
+To update login ID, API consumer should instead perform a series of operations:
+- Create password authenticated principal with new login ID.
+- Ensure new principal is usable. (e.g. verified email/phone, approved by admin)
+- Delete the principal of old login ID.
+
+For passcodes (#293), a new principal associated with a specical identity
+provider should be used instead.
+
+The principal manipulation APIs would be specified in another specification.
+
+This address use case 1, 6.
+
+## Allow Multiple Login ID per Login ID Key
+In some website, user can associate multiple emails (potentially unlimited) with
+same user. These email may be used in auth features:
+- Login with multiple emails
+- Backup email
+- Welcome email
+
+A login ID key should allow having multiple associated login ID.
+When signing up, user should provide a list of login ID. e.g.
+```json
+{
+    "login_ids": [
+        { "username": "kiootic" },
+        { "email": "kiootic+1@oursky.com" },
+        { "email": "kiootic+2@oursky.com" }
+    ],
+    "password": "12345678"
+}
+```
+
+This addresses use case 6.
+
+
+## Standard Semantics Keys
```

Comment:
How does the semantics keys relating with metadata? I suppose `username`, `email`, `phone` are login ID keys and it is okay to be in `principle`. However `name`, `picture`, `locale`, `zoneinfo` etc shall be in metadata of user object, and it seems redundant to be in principle. (Or what is principle actually?)